cql3/Cql.g

```diff
@@ -1422,74 +1423,81 @@ idxName[cql3::index_name& name]
-    @init{std::vector<std::pair<::shared_ptr<cql3::term::raw>, ::shared_ptr<cql3::term::raw>>> m;}
+    : t=STRING_LITERAL {
+                    // This is a workaround for antlr3 not distinguishing between
+                    // calling in lexer setText() with an empty string and not calling
+                    // setText() at all.
+                    auto text = $t.text;
+                    if (text.size() == 1 && text[0] == '\xFF') {
+                        text = {};
+                    }
+                    $constant = cql3::expr::untyped_constant{cql3::expr::untyped_constant::string, std::move(text)}; }
+    | t=INTEGER        { $constant = cql3::expr::untyped_constant{cql3::expr::untyped_constant::integer, $t.text}; }
+    | t=FLOAT          { $constant = cql3::expr::untyped_constant{cql3::expr::untyped_constant::floating_point, $t.text}; }
+    | t=BOOLEAN        { $constant = cql3::expr::untyped_constant{cql3::expr::untyped_constant::boolean, $t.text}; }
+    | t=DURATION       { $constant = cql3::expr::untyped_constant{cql3::expr::untyped_constant::duration, $t.text}; }
+    | t=UUID           { $constant = cql3::expr::untyped_constant{cql3::expr::untyped_constant::uuid, $t.text}; }
+    | t=HEXNUMBER      { $constant = cql3::expr::untyped_constant{cql3::expr::untyped_constant::hex, $t.text}; }
+    | { sign=""; } ('-' {sign = "-"; } )? t=(K_NAN | K_INFINITY) { $constant = cql3::expr::untyped_constant{cql3::expr::untyped_constant::floating_point, sign + $t.text}; }
+    ;
+
+mapLiteral returns [cql3::expr::expression map]
+    @init{std::vector<cql3::expr::expression> m;}
     : '{' { }
-          ( k1=term ':' v1=term { m.push_back(std::pair<shared_ptr<cql3::term::raw>, shared_ptr<cql3::term::raw>>{k1, v1}); } ( ',' kn=term ':' vn=term { m.push_back(std::pair<shared_ptr<cql3::term::raw>, shared_ptr<cql3::term::raw>>{kn, vn}); } )* )?
-      '}' { $map = ::make_shared<cql3::maps::literal>(m); }
+          ( k1=term ':' v1=term { m.push_back(cql3::expr::tuple_constructor{{std::move(k1), std::move(v1)}}); }
+            ( ',' kn=term ':' vn=term { m.push_back(cql3::expr::tuple_constructor{{std::move(kn), std::move(vn)}}); } )*  )?
+      '}' { $map = cql3::expr::collection_constructor{cql3::expr::collection_constructor::style_type::map, std::move(m)}; }
     ;
 
-setOrMapLiteral[shared_ptr<cql3::term::raw> t] returns [shared_ptr<cql3::term::raw> value]
-	@init{ std::vector<std::pair<shared_ptr<cql3::term::raw>, shared_ptr<cql3::term::raw>>> m;
-	       std::vector<shared_ptr<cql3::term::raw>> s;
-	}
-    : ':' v=term { m.push_back({t, v}); }
-          ( ',' kn=term ':' vn=term { m.push_back({kn, vn}); } )*
-      { $value = ::make_shared<cql3::maps::literal>(std::move(m)); }
-    | { s.push_back(t); }
-          ( ',' tn=term { s.push_back(tn); } )*
-      { $value = ::make_shared<cql3::sets::literal>(std::move(s)); }
+setOrMapLiteral[cql3::expr::expression t] returns [cql3::expr::expression value]
```

Comment:
This return type implies that a setOrMapLiteral might be a binary operation, conjunction, or a bool, which isn't quite true.  Why not change it to collection_constructor (the lowest common ancestor node for set and map)?

Reply:
I think it reads better and is more future-proof if setOrMapLiteral cannot be a binop.

Reply:
Done.